frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -538,6 +549,286 @@ describe("QueryBuilder", () => {
+              `${TEST_MODEL_DATASET_COLUMN.display_name}X`,
+            );
+          });
+
+          userEvent.tab();
+
+          await waitFor(() => {
+            expect(
+              screen.getByRole("button", { name: "Save changes" }),
+            ).toBeEnabled();
+          });
+
+          history.goBack();
+
+          expect(
+            screen.getByText("Changes were not saved"),
+          ).toBeInTheDocument();
+          expect(
+            screen.getByText(
+              "Navigating away from here will cause you to lose any changes you have made.",
+            ),
+          ).toBeInTheDocument();
+        });
+
+        it("does not show custom warning modal when leaving unedited metadata via SPA navigation", async () => {
+          const { history } = await setup({
+            card: TEST_MODEL_CARD,
+            dataset: TEST_MODEL_DATASET,
+            initialRoute: "/home",
+          });
+
+          history.push(`/model/${TEST_MODEL_CARD.id}/metadata`);
+
+          await waitFor(() => {
+            expect(
+              screen.queryByTestId("loading-spinner"),
+            ).not.toBeInTheDocument();
+          });
+
+          const columnDisplayName = await screen.findByTitle("Display name");
+
+          userEvent.click(columnDisplayName);
+          userEvent.type(columnDisplayName, "X");
+
+          await waitFor(() => {
+            expect(columnDisplayName).toHaveValue(
+              `${TEST_MODEL_DATASET_COLUMN.display_name}X`,
+            );
+          });
+
+          userEvent.tab();
+
+          userEvent.click(columnDisplayName);
+          userEvent.type(columnDisplayName, "{backspace}");
+
+          await waitFor(() => {
+            expect(columnDisplayName).toHaveValue(
+              TEST_MODEL_DATASET_COLUMN.display_name,
+            );
+          });
+
+          userEvent.tab();
+
+          await waitFor(() => {
+            expect(
+              screen.getByRole("button", { name: "Save changes" }),
+            ).toBeDisabled();
+          });
+
+          history.goBack();
+
+          expect(
+            screen.queryByText("Changes were not saved"),
+          ).not.toBeInTheDocument();
+          expect(
+            screen.queryByText(
+              "Navigating away from here will cause you to lose any changes you have made.",
+            ),
+          ).not.toBeInTheDocument();
+        });
+
+        it("does not show custom warning modal when saving edited metadata", async () => {
+          const { history } = await setup({
+            card: TEST_MODEL_CARD,
+            dataset: TEST_MODEL_DATASET,
+            initialRoute: `/model/${TEST_MODEL_CARD.id}/query`,
+          });
+
+          /**
+           * When initialRoute is `/model/${TEST_MODEL_CARD.id}/metadata`,
+           * the QueryBuilder gets incompletely intialized.
+           * This seems to affect only tests.
+           */
+          userEvent.click(screen.getByText("Metadata"));
```

Comment:
I spent over an hour debugging this and could not figure it out. I'd appreciate any suggestions on what could be wrong here.

This is how the code should look like:

```suggestion
            initialRoute: `/model/${TEST_MODEL_CARD.id}/metadata`,
          });
```